app/services/types_de_champ_service.rb

```diff
@@ -22,6 +24,33 @@ def self.create_update_procedure_params(params, private=false)
 
   private
 
+  def self.order_champs(params)
+    tdcas = params[:procedure][:types_de_champ_attributes].to_a
+
+    tdcas
+      .select { |tdca| !is_number?(tdca[1][:custom_order_place]) }
+      .each { |tdca| tdca[1][:custom_order_place] = (tdca[1][:order_place].to_i + 1).to_s }
+
+    out_of_order_tdcas, ordered_tdcas = tdcas.partition { |tdca| (tdca[1][:order_place].to_i + 1) != tdca[1][:custom_order_place].to_i }
+
+    # out_of_order_tdcas are in first positions
+    # to win the right position in the next sort_by
+    tdcas = out_of_order_tdcas + ordered_tdcas
+
+    ordered_tdcas = tdcas
+      .sort_by { |el| el[1][:custom_order_place].to_i }
+      .each_with_index { |el, index| el[1][:order_place] = index.to_s }
+      .reduce({}) { |acc, el| acc[el[0]] = el[1]; acc }
+
+    params[:procedure][:types_de_champ_attributes] = ActionController::Parameters.new(ordered_tdcas)
+
+    params
+  end
+
+  def self.is_number?(value)
+    (value =~ /^[0-9]+$/) == 0
```

Comment:
Est ce que `value == value.to_i.to_s` serait plus lisible ?
https://stackoverflow.com/questions/2095493/how-can-i-check-if-a-value-is-a-number